cpp/src/arrow/filesystem/s3fs.h

```diff
@@ -69,6 +69,13 @@ enum class S3CredentialsKind : int8_t {
   WebIdentity
 };
 
+/// Pure virtual class for describing custom S3 retry strategies
+class S3RetryStrategy {
+ public:
+  virtual bool ShouldRetry(Status& error, long attempted_retries) = 0;
+  virtual long CalculateDelayBeforeNextRetry(Status& error, long attempted_retries) = 0;
```

Comment:
The `Status error` parameter itself, isn't very useful, since all S3 errors seem to result in the same IOError code. Having access to the AWS/S3 error code here would be really useful.
Any objections to adding that as a third argument to these methods? It could be casted as an int to avoid having to include any aws headers here.

Reply:
That could work, and/or it could be provided via a StatusDetail though that may not be as simple. https://github.com/apache/arrow/blob/72b52ef1b2681f927e8cfc2bc6a643648c542058/cpp/src/arrow/status.h#L99-L113

Reply:
I opted to pass around a struct containing the error code and a couple of other fields from the AWS error object. I looked into StatusDetail, but I don't think it'll work in this case since there's several fields that are important in deciding whether or not something should be retried (For example, `ConnectRetryStrategy` uses both `ShouldRetry()` and `GetExceptionName()`)

Reply:
Ah, the intent of StatusDetail is that you subclass it and add all the fields you want, then you can cast it back where needed. Example: https://github.com/apache/arrow/blob/b1f90c48bc39452c9c236a988b4bbd38fea96bae/cpp/src/arrow/flight/types.h#L81-L109

It is definitely more work to do things that way. However it may be worth doing if you think callers outside of the retry API would care about these error details as well.

Reply:
That said I don't feel too strongly about this. This can always be converted into a StatusDetail in the future and appended to errors if necessary.